File: model/estudiantes.py
from connection.connection import Conecction


class estudiantes:
    def buscarEstudiantes():
        query = "SELECT idestudiante, idcurso, nombres, apellido_p , apellido_m, rut, telefono, direccion FROM estudiante"
        tipoConsulta = 2
        conexionBD = Conecction()
        conexionBD.conectar()
        resultado = conexionBD.consultaDB(query, tipoConsulta)
        conexionBD.desconectar()
        return resultado

    def buscarEstudiante(id):
        query = "SELECT  nombres, apellido_p, apellido_m, rut, telefono, direccion FROM estudiante WHERE idestudiante = %s"
        parametros = (id,)
        tipoConsulta = 2
        conexionBD = Conecction()
        conexionBD.conectar()
        resultado = conexionBD.consultaDB(query, tipoConsulta, parametros)
        conexionBD.desconectar()
        return resultado

    def buscarEstudianteYApoderados(id):
        query = "select e.idestudiante, e.idcurso, e.nombres as estudiante_nombre, e.apellido_p as estudiante_apellido_p, e.apellido_m as estudiante_apellido_m, e.rut as estudiante_rut, e.telefono as estudiante_telefono, e.direccion as estudiante_direccion, e.idapoderado, e.idapoderadosup, e.avatar, a.nombres as apoderado_nombres, a.apellido_p as apoderado_apellido_p, a.apellido_m as apoderado_apellido_m, a.telefono as apoderado_telefono, a.rut as apoderado_rut, asup.nombres as apoderadosup_nombres, asup.apellido_p as apoderadosup_apellido_p, asup.apellido_m as idapoderadosup_apellido_m, asup.telefono as apoderadosup_telefono, asup.rut as apoderadosup_rut from estudiante e left join apoderado a on e.idapoderado = a.idapoderado left join apoderadosup asup on e.idapoderadosup = asup.idapoderadosup where idestudiante = %s"
        parametros = (id,)
        tipoConsulta = 2
        conexionBD = Conecction()
        conexionBD.conectar()
        resultado = conexionBD.consultaDB(query, tipoConsulta, parametros)
        conexionBD.desconectar()
        return resultado

    def buscarApoderadoPorRut(rut):
        query = "select idapoderado from apoderado where rut = %s"
        parametros = (rut,)
        tipoConsulta = 2
        conexionBD = Conecction()
        conexionBD.conectar()
        resultado = conexionBD.consultaDB(query, tipoConsulta, parametros)
        conexionBD.desconectar()
        return resultado

    def buscarApoderadoSupPorRut(rut):
        query = "select idapoderadosup from apoderadosup where rut = %s"
        parametros = (rut,)
        tipoConsulta = 2
        conexionBD = Conecction()
        conexionBD.conectar()
        resultado = conexionBD.consultaDB(query, tipoConsulta, parametros)
        conexionBD.desconectar()
        return resultado

    def crearEstudiante(
        nombres,
        idcurso,
        apellido_p,
        apellido_m,
        rut,
        telefono,
        direccion,
        idapoderado,
        idapoderadosup,
        avatar,
    ):
        query = "INSERT INTO estudiante (nombres, idcurso, apellido_p,apellido_m,rut,telefono,direccion, idapoderado, idapoderadosup, avatar) VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s);"
        tipoConsulta = 1
        parametros = (
            nombres,
            idcurso,
            apellido_p,
            apellido_m,
            rut,
            telefono,
            direccion,
            idapoderado,
            idapoderadosup,
            avatar,
        )
        conexionBD = Conecction()
        conexionBD.conectar()
        try:
            conexionBD.consultaDB(query, tipoConsulta, parametros)
            print("Se ha insertado Correctamente")
        except Exception as error:
            print("Ha ocurrido un problema en la inserción", error)
        conexionBD.desconectar()

    def editarEstudiante(
        nombres,
        idcurso,
        apellido_p,
        apellido_m,
        rut,
        telefono,
        direccion,
        idapoderado,
        idapoderadosup,
        avatar,
        idestudiante,
    ):
        query = "UPDATE befast.estudiante SET idcurso=%s, nombres=%s, apellido_p=%s, apellido_m=%s, rut=%s, telefono=%s, direccion=%s, idapoderado=%s, idapoderadosup=%s, avatar=%s WHERE idestudiante=%s;"
        tipoConsulta = 1
        parametros = (
            idcurso,
            nombres,
            apellido_p,
            apellido_m,
            rut,
            telefono,
            direccion,
            idapoderado,
            idapoderadosup,
            avatar,
            idestudiante,
        )
        conexionBD = Conecction()
        conexionBD.conectar()
        try:
            conexionBD.consultaDB(query, tipoConsulta, parametros)
            print("Se ha Editado Correctamente")
        except Exception as error:
            print("Ha ocurrido un problema en la edición", error)
        conexionBD.desconectar()

    def eliminarEstudiante(id):
        query = "DELETE FROM estudiante WHERE idestudiante = %s;"
        tipoConsulta = 1
        parametros = (id,)
        conexionBD = Conecction()
        conexionBD.conectar()
        try:
            conexionBD.consultaDB(query, tipoConsulta, parametros)
            print("Se ha Eliminado Correctamente")
        except:
            print("Ha ocurrido un problema en la Eliminación")
        conexionBD.desconectar()

File: view/js/estudiantes.js
function updateTable() {
  $.ajax({
    url: "http://localhost:8000/api/estudiantes",
    type: "GET",
    contentType: "application/json ; charset=utf8",
    dataType: "json",
    success: function (response) {
      var tabla = $("#myTable > tbody:last-child");
      tabla.empty();
      response.forEach((estudiantes) => {
        var id = estudiantes.id;
        var idcurso = estudiantes.idcurso;
        var nombres = estudiantes.nombres;
        var apellido_p = estudiantes.apellido_p;
        var apellido_m = estudiantes.apellido_m;
        var rut = estudiantes.rut;
        var telefono = estudiantes.telefono;
        var direccion = estudiantes.direccion;
        var fila =
          '<tr class="fila">' +
          '<td class="id">' +
          id +
          "</td>" +
          "<td>" +
          idcurso +
          "</td>" +
          "<td>" +
          nombres +
          "</td>" +
          "<td>" +
          apellido_p +
          "</td>" +
          '<td><button class="btn btn-success btn-ver" onclick="btnVer()"><i class="fas fa-eye"></i> Ver más</button></td>' +
          "<tr>";
        tabla.append(fila);
      });
    },
  });
}

updateTable();

$(document).ready(() => {
  $.ajax({
    url: "http://127.0.0.1:8000/api/cursos/",
    type: "GET",
    contentType: "application/json ; charset=utf8",
    dataType: "json",
    success: function (response) {
      var opt = $("#idcurso");

      response.forEach((cursos) => {
        var id = cursos.idcurso;
        var nivel = cursos.nivel;
        var letra = cursos.letra;
        var opcion =
          '<option value="' + id + '">' + nivel + letra + "</option>";
        console.log(opcion);
        opt.append(opcion);
      });
    },
  });
});

$("#crearEstudianteBtn").click(async () => {
  var nombres = $("#nombres").val();
  var idcurso = $("#idcurso").val();
  var apellido_p = $("#apellido_p").val();
  var apellido_m = $("#apellido_m").val();
  var rut = $("#rutcrear").val();
  var telefono = $("#telefono").val();
  var direccion = $("#direccion").val();

  async function buscarRUT(sup, rut) {
    var url = sup
      ? `http://localhost:8000/api/apoderadosuprut/${rut}`
      : `http://localhost:8000/api/apoderadorut/${rut}`;

    console.log(rut)
    console.log(url);

    try {
      var response = await $.ajax({
        url: url,
        type: "GET",
        contentType: "application/json; charset=utf8",
        dataType: "json",
      });

      // Verifica si es apoderado (sup === false) para obtener idapoderado
      if (sup === false) {
        return response.idapoderado; // Cambia a idapoderado
      } else {
        return response.idapoderadosup; // Mantén idapoderadosup
      }
    } catch (error) {
      throw error;
    }
  }

  try {
    var idapoderadosup = await buscarRUT(
      true,
      parseInt(document.getElementById("rutapoderadosup").value)
    );
    var idapoderado = await buscarRUT(
      false,
      parseInt(document.getElementById("rutapoderado").value)
    );
    var avatar = $("#avatarestudiante").val();
    console.log(avatar, idapoderado, idapoderadosup);

    var datos = {
      nombres: nombres,
      idcurso: idcurso,
      apellido_p: apellido_p,
      apellido_m: apellido_m,
      rut: rut,
      telefono: telefono,
      direccion: direccion,
      idapoderado: idapoderado,
      idapoderadosup: idapoderadosup,
      avatar: avatar,
    };

    console.log(datos)
    
    var response = await $.ajax({
      url: "http://localhost:8000/api/estudiante/",
      type: "POST",
      contentType: "application/json; charset=utf8",
      data: JSON.stringify(datos),
    });

    console.log(response);

    updateTable();
  } catch (error) {
    console.error(error);
  }
});

function btnVer() {
  var button = event.target;
  var row = button.closest("tr");
  var idElement = row.querySelector(".id");
  var id = idElement.textContent;

  $.ajax({
    url: `http://localhost:8000/api/estudianteyapoderado/${id}`,
    type: "GET",
    contentType: "application/json ; charset=utf8",
    dataType: "json",
    success: (response) => {
      localStorage.setItem("idestudiante", response.idestudiante);
      localStorage.setItem("idapoderado", response.idapoderado);
      localStorage.setItem("idapoderadosup", response.idapoderadosup);

      $(".ficha").removeClass("hide");
      function updateFicha() {
        $("#nombre-estudiante").html(
          response.estudiante_nombre +
            " " +
            response.estudiante_apellido_p +
            " " +
            response.estudiante_apellido_m
        );
        $("#rut-estudiante").html(response.rut);
        $("#contacto-estudiante").html(response.telefono);

        // Apoderados
        $("#nombre-apoderado-principal").html(
          response.apoderado_nombres +
            " " +
            response.apoderado_apellido_p +
            " " +
            response.apoderado_apellido_m
        );

        $("#telefono-apoderado-principal").html(
          "+56" + response.apoderado_telefono
        );

        $("#nombre-apoderado-suplente").html(
          response.apoderadosup_nombres +
            " " +
            response.apoderadosup_apellido_p +
            " " +
            response.apoderadosup_apellido_m
        );
        $("#telefono-apoderado-suplente").html(
          "+56" + response.apoderadosup_telefono
        );

        $("#avatar").attr("src", response.avatar);
      }
      updateFicha();
    },
  });
}

function btnEditarModal() {
  var id = localStorage.getItem("idestudiante");
  $("#modalEditar").modal("show");
  function datosModal() {
    $.ajax({
      url: `http://localhost:8000/api/estudianteyapoderado/${id}`,
      type: "GET",
      contentType: "application/json ; charset=utf8",
      dataType: "json",
      success: (response) => {
        $("#ed_nombres").val(response.estudiante_nombre);
        $("#ed_idcurso").val(response.idcurso);
        $("#ed_apellido_p").val(response.estudiante_apellido_p);
        $("#ed_apellido_m").val(response.estudiante_apellido_m);
        $("#ed_rutcrear").val(response.estudiante_rut);
        $("#ed_telefono").val(response.estudiante_telefono);
        $("#ed_direccion").val(response.estudiante_direccion);
        $("#ed_rutapoderado").val(response.apoderado_rut);
        $("#ed_rutapoderadosup").val(response.apoderadosup_rut);
        $("#ed_avatarestudiante").val(response.avatar);
      },
    });
  }

  datosModal();

  $("#editarBtn").click(async () => {
    var nombres = $("#ed_nombres").val();
    var idcurso = $("#ed_idcurso").val();
    var apellido_p = $("#ed_apellido_p").val();
    var apellido_m = $("#ed_apellido_m").val();
    var rut = $("#ed_rutcrear").val();
    var telefono = $("#ed_telefono").val();
    var direccion = $("#ed_direccion").val();

    async function buscarRUT(sup, rut) {
      var url = sup
        ? `http://localhost:8000/api/apoderadosup/${rut}`
        : `http://localhost:8000/api/apoderado/${rut}`;

      console.log(url);

      try {
        var response = await $.ajax({
          url: url,
          type: "GET",
          contentType: "application/json; charset=utf8",
          dataType: "json",
        });

        // Verifica si es apoderado (sup === false) para obtener idapoderado
        if (sup === false) {
          return response.idapoderado; // Cambia a idapoderado
        } else {
          return response.idapoderadosup; // Mantén idapoderadosup
        }
      } catch (error) {
        throw error;
      }
    }

    try {
      var idapoderadosup = await buscarRUT(
        true,
        parseInt($("#ed_rutapoderadosup").val())
      );
      var idapoderado = await buscarRUT(
        false,
        parseInt($("#ed_rutapoderado").val())
      );
      var avatar = $("#ed_avatarestudiante").val();
      console.log(avatar);

      var datos = {
        nombres: nombres,
        idcurso: idcurso,
        apellido_p: apellido_p,
        apellido_m: apellido_m,
        rut: rut,
        telefono: telefono,
        direccion: direccion,
        idapoderado: idapoderado,
        idapoderadosup: idapoderadosup,
        avatar: avatar,
        id: id,
      };
      var response = await $.ajax({
        url: `http://localhost:8000/api/estudiante/${id}`,
        type: "PUT",
        contentType: "application/json; charset=utf8",
        data: JSON.stringify(datos),
      });

      console.log(response);

      updateTable();
    } catch (error) {
      console.error(error);
    }
  });

  function btnBorrar() {
    var id = localStorage.getItem("idestudiante");

    var datos = {
      id: id,
    };

    $.ajax({
      url: `http://127.0.0.1:8000/api/estudiante/${id}`,
      type: "DELETE",
      data: JSON.stringify(datos),
      contentType: "application/json ; charset=utf8",
      dataType: "json",
      success: (response) => {
        updateTable();
      },
    });
  }
}
